versioned/persist/adapter/src/main/java/org/projectnessie/versioned/persist/adapter/spi/AbstractDatabaseAdapter.java

```diff
@@ -626,14 +629,33 @@ protected void validateHashExists(OP_CONTEXT ctx, Hash hash) throws ReferenceNot
   }
 
   /** Load the commit-log entry for the given hash, return {@code null}, if not found. */
-  protected abstract CommitLogEntry fetchFromCommitLog(OP_CONTEXT ctx, Hash hash);
+  protected final CommitLogEntry fetchFromCommitLog(OP_CONTEXT ctx, Hash hash) {
+    try (Traced ignore = trace("fetchFromCommitLog").tag(TAG_HASH, hash.asString())) {
+      return doFetchFromCommitLog(ctx, hash);
+    }
+  }
+
+  protected abstract CommitLogEntry doFetchFromCommitLog(OP_CONTEXT ctx, Hash hash);
 
   /**
    * Fetch multiple {@link CommitLogEntry commit-log-entries} from the commit-log. The returned list
    * must have exactly as many elements as in the parameter {@code hashes}. Non-existing hashes are
    * returned as {@code null}.
    */
-  protected abstract List<CommitLogEntry> fetchPageFromCommitLog(OP_CONTEXT ctx, List<Hash> hashes);
+  protected final List<CommitLogEntry> fetchPageFromCommitLog(OP_CONTEXT ctx, List<Hash> hashes) {
+    if (hashes.isEmpty()) {
+      return Collections.emptyList();
+    }
+    try (Traced ignore =
```

Comment:
we're leaking tracing code into the existing implementation or am I missing something?

Reply:
It's tracing about the inner workings of the implementation, so yes.

Reply:
what does this mean if tracing is not enabled? Is this a no-op or are we creating unnecessary objects?

Reply:
noop

Reply:
I think it's still unclear to me from reading the code why this would be a noop when tracing is disabled. `Traced.trace()` creates & starts a span so it's still adding object creation overhead or not? Maybe this can be clarified with a comment or so in `Traced`?

Reply:
It's still a Noop implementation in OpenTracing, right?

Reply:
I haven't looked into OpenTracing, that's why I'm asking, because it's unclear to people reading the code when not knowing implementation details of OpenTracing itself

Reply:
If disabled Open Tracing will return non-null noop implementations, but it looks like we will still wrap them in another layer of `Traced` objects.

From my POV it is fine, unless we want to optimize for the disabled tracing use case, which I guess we do not actually want since we expect tracing to be used with higher or lower degree of sampling depending on the deployment.